go/vt/vtgate/vstream_manager.go

```diff
@@ -676,6 +696,24 @@ func (vs *vstream) sendAll(ctx context.Context, sgtid *binlogdatapb.ShardGtid, e
 	return nil
 }
 
+// generateFullyCopyCompleted returns a new copy_completed event only if all stream has received a copy_completed event.
+// This new event represents the completion of all the copy operations.
+func (vs *vstream) generateFullyCopyCompleted(ctx context.Context, sgtid *binlogdatapb.ShardGtid, event *binlogdatapb.VEvent) *binlogdatapb.VEvent {
+	vs.mu.Lock()
+	defer vs.mu.Unlock()
+
+	vs.copyCompletedShard[fmt.Sprintf("%s/%s", event.Keyspace, event.Shard)] = struct{}{}
+
+	for _, shard := range vs.vgtid.ShardGtids {
+		if _, ok := vs.copyCompletedShard[fmt.Sprintf("%s/%s", shard.Keyspace, shard.Shard)]; !ok {
+			return nil
+		}
+	}
+	return &binlogdatapb.VEvent{
+		Type: binlogdatapb.VEventType_COPY_COMPLETED,
```

Comment:
This event has neither keyspace nor shard.
It represents the completion of data copy wherever the data is.